bundles/test/org.eclipse.smarthome.magic/src/main/java/org/eclipse/smarthome/magic/binding/internal/MagicHandlerFactory.java

```diff
@@ -113,14 +125,40 @@ protected ThingHandler createHandler(Thing thing) {
         if (thingTypeUID.equals(THING_TYPE_IMAGE)) {
             return new MagicImageHandler(thing);
         }
+        if (thingTypeUID.equals(THING_TYPE_ACTION_MODULE)) {
+            MagicActionModuleThingHandler handler = new MagicActionModuleThingHandler(thing);
+            MagicThingActionsService actionProvider = new MagicThingActionsService(handler);
+
+            Dictionary<String, String> properties = new Hashtable<>();
+            properties.put("esh.thingActionID", thing.getUID().getAsString());
+
+            this.actionServiceRegs.put(thing.getUID(),
+                    this.bundleContext.registerService(AnnotatedActions.class, actionProvider, properties));
+
+            return handler;
+        }
         if (thingTypeUID.equals(THING_TYPE_ONLINE_OFFLINE)) {
             return new MagicOnlineOfflineHandler(thing);
         }
-
         if (thingTypeUID.equals(THING_TYPE_BRIDGE_1) || thingTypeUID.equals(THING_TYPE_BRIDGE_2)) {
             return new MagicBridgeHandler((Bridge) thing);
         }
 
         return null;
     }
+
+    @Override
+    protected void removeHandler(ThingHandler thingHandler) {
+        if (thingHandler instanceof MagicActionModuleThingHandler) {
+            ServiceRegistration<?> serviceReg = this.actionServiceRegs.remove(thingHandler.getThing().getUID());
+            if (serviceReg != null) {
+                // remove actionprovider service, if handler is removed
+                AnnotatedActions service = (AnnotatedActions) bundleContext.getService(serviceReg.getReference());
+                serviceReg.unregister();
+                // if (service != null) {
+                // service.deactivate();
```

Comment:
Can this lines be removed? Does the framework call the deactivate or not?

Reply:
The framework now takes care of registering and unregistering of services.